server/rpc.go

```
@@ -0,0 +1,462 @@
+package server
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"math/big"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	"github.com/livepeer/go-livepeer/core"
+	"github.com/livepeer/go-livepeer/eth"
+	lpTypes "github.com/livepeer/go-livepeer/eth/types"
+	"github.com/livepeer/lpms/stream"
+
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/proto"
+)
+
+var AuthType_LPE = "Livepeer-Eth-1"
+
+type orchestrator struct {
+	transcoder string
+	address    ethcommon.Address
+	node       *core.LivepeerNode
+}
+
+type Orchestrator interface {
+	Transcoder() string
+	Address() ethcommon.Address
+	Sign([]byte) ([]byte, error)
+	CurrentBlock() *big.Int
+	GetJob(int64) (*lpTypes.Job, error)
+	TranscodeSeg(*lpTypes.Job, *core.SignedSegment) error
+	StreamIDs(*lpTypes.Job) ([]core.StreamID, error)
+}
+
+// Orchestator interface methods
+func (orch *orchestrator) Transcoder() string {
+	return orch.transcoder
+}
+
+func (orch *orchestrator) CurrentBlock() *big.Int {
+	if orch.node == nil || orch.node.Eth == nil {
+		return nil
+	}
+	block, _ := orch.node.Eth.LatestBlockNum()
```

Check error returned here? If `block` is nil could get some nil pointer exceptions

---

Actually that's okay to ignore the error here because we return *big.Int only (and not an error) but none of the callers of `CurrentBlock()` actually check the result, so added that in...

The only caller was the `blockInRange` helper. Returned `true` if the block was nil; I figure it's better to leave the benefit of doubt here. If the transcoder has problems fetching the block, then it needs to get those problems fixed; we shouldn't compromise the broadcaster's job because of an unrelated, internal issue in checking validity. Totally open to discussion.

---

Ah because `orch.node.Eth.LatestBlockNum()` uses a constant backoff with no max # of retries so in theory if it returns something it will always be a valid `*big.Int` (I actually don't think it should ever really return nil except for maybe the case where you're talking to a node with 0 blocks) or else it would just keep continuously retrying...perhaps later on the function return signature could be updated, don't think its necessary here though

> If the transcoder has problems fetching the block, then it needs to get those problems fixed; we shouldn't compromise the broadcaster's job because of an unrelated, internal issue in checking validity. 

I think that is fine for now, but in the future at a more stable point I think there is an argument for stopping everything you're doing if you are unable to perform a validity check properly

---

We subscribe to new block events within the JobService and write it to disk. This was done for job recovery, but we don't need to recover jobs anymore. The subscription could still be useful on its own though.

Rather than look up the block number via eth RPC every time, I was actually thinking it might be better to look up the block number from the database or even store it in a global var. Faster, more reliable, and less unnecessary load on the network. This is especially relevant here, since the lookup is being done during the transcoding flow, which is under tighter time constraints; don't want a geth stall to back up the transcoding.